Make remove_virtual_function() consistent

owen-mc-diffblue opened the pull request:
If there is more than one possible virtual function, we cast the object
to the appropriate type in each case, but if there is only one
possibility we don't. This PR makes us cast the object when there is
only one possibility as well.

Note that just because there is only one possibility, it does not mean
that the object already has the correct type. Also, if the object is
already the correct type then the cast isn't inserted, in all cases.

Commit:
Make remove_virtual_function() consistent

If there is more than one possible virtual function, we cast the object
to the appropriate type in each case, but if there is only one
possibility we don't. This PR makes us cast the object when there is
only one possibility as well.

Note that just because there is only one possibility, it does not mean
that the object already has the correct type. Also, if the object is
already the correct type then the cast isn't inserted, in all cases.

tautschnig commented:
@owen-jones-diffblue Would it be possible to add a regression test, in particular one that fits with "Note that just because there is only one possibility, it does not mean that the object already has the correct type."?

owen-mc-diffblue commented:
Sure. I will try to get around to it soon.

Commit:
Test casts in remove_virtual_function()

Add a test to check that when we replace virtual function calls with
static function calls we cast the class to the appropriate type for the
static function call, so the argument for the `this` parameter will
have the correct type, and also that when a cast is not needed we will
not create one. Different code is run if there is only one possible
static function or if there is more than one, so we make sure that we
test both of these options.

owen-mc-diffblue commented:
I've just added a test, which I've verified fails on develop but passes on this branch.

owen-jones-diffblue commented:
This PR is ready to be merged

tautschnig commented on regression/cbmc-java/remove_virtual_function_typecast/VirtualFunctions.java:
```diff
@@ -0,0 +1,27 @@
+    // single possibility, needs cast
+    c.f();
+  }
+}
```
GitHub says there is no newline at end-of-file.

tautschnig reviewed (commented):
Minor comment about a newline, but more importantly: could you please extend the commit message such that it is clear even without the context of this particular PR? It's not as easy to see in the git history later on what this test is actually about.
